### PR TITLE
Add C# API documentation for training (and some other changes)

### PR DESCRIPTION
**Pull request by baijumeswani**
This pull request:

- adds the C# API documentation for training
- changes the name from learning on the edge to on-device training for existing python and C/C++ API (will require gh-pages changes before a preview can be generated since it collides with #15900).
- reverts changes in #15854 (back to using ubuntu now).

C# documentation preview at https://baijumeswani.github.io/onnxruntime/docs/api/csharp/api/

### CODE REVIEW

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs [hunk 1]
```diff
@@ -7,7 +7,16 @@ namespace Microsoft.ML.OnnxRuntime
...
-    ///  Holds the Checkpoint State as generated/consumed by on-device training APIs
+    ///  Holds the state of the training session.
+    /// This class holds the entire training session state that includes model parameters, their gradients,
+    ///optimizer parameters, and user properties. The TrainingSession leverages the CheckpointState
+    /// by accessing and updating the contained training state.
+    /// <note type="note">
+    /// Note that the training session created with a checkpoint state uses this state to store the entire
+    /// training state (including model parameters, its gradients, the optimizer states and the properties).
+    /// The TrainingSession does not hold a copy of the CheckpointState and as a result, it is required
+    /// that the checkpoint state outlive the lifetime of the training session.
```
Comment (pengwa) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs:18]:
```suggestion
    /// that the checkpoint state outlives the lifetime of the training session.
```

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs [hunk 2]
```diff
@@ -7,7 +7,16 @@ namespace Microsoft.ML.OnnxRuntime
 {
 #if __ENABLE_TRAINING_APIS__
     /// <summary>
-    ///  Holds the Checkpoint State as generated/consumed by on-device training APIs
+    ///  Holds the state of the training session.
+    /// This class holds the entire training session state that includes model parameters, their gradients,
+    ///optimizer parameters, and user properties. The TrainingSession leverages the CheckpointState
```
Comment (pengwa) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs:12]:
```suggestion
    /// optimizer parameters, and user properties. The TrainingSession leverages the CheckpointState
```

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs [hunk 3]
```diff
@@ -55,9 +64,15 @@ private void AddPropertyImpl<T>(string propertyName, PropertyType propertyType,
         public override bool IsInvalid { get { return handle == IntPtr.Zero; } }
 
         /// <summary>
-        /// Loads Checkpoint state from path
+        /// Load a checkpoint state from directory on disk into checkpoint_state.
```
Comment (pengwa) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs:67]:
```suggestion
        /// Load a checkpoint state from a directory on disk into checkpoint_state.
```

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs [hunk 4]
```diff
@@ -33,7 +33,16 @@ enum LRScheduler
     }
 
     /// <summary>
-    /// Represents a Training Session on an ONNX Model.
+    /// Trainer class that provides training, evaluation and optimizer methods for training an ONNX models.
```
Comment (pengwa) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs:36]:
```suggestion
    /// Trainer class that provides training, evaluation and optimizer methods for training an ONNX model.
```

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs [hunk 5]
```diff
@@ -33,7 +33,16 @@ enum LRScheduler
...
+    ///
+    /// The training session requires four training artifacts
+    /// - The training onnx model
+    /// - The evaluation onnx model (optional)
+    /// - The optimizer onnx model
+    /// - The checkpoint directory
+    ///
+    /// These artifacts can be generated using the `onnxruntime-training` python [utility](https://github.com/microsoft/onnxruntime/blob/main/orttraining/orttraining/python/training/onnxblock/README.md).
+    ///
     /// This is a IDisposable class and it must be disposed of
```
Comment (pengwa) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs:46]:
```suggestion
    /// This is an IDisposable class and it must be disposed of
```

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs [hunk 6]
```diff
@@ -33,7 +33,16 @@ enum LRScheduler
...
     /// using either a explicit call to Dispose() method or
```
Comment (pengwa) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs:47]:
```suggestion
    /// using either an explicit call to Dispose() method or
```

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs [hunk 7]
```diff
@@ -33,7 +33,16 @@ enum LRScheduler
...
     /// a pattern of using() block. If this is a member of another
```
Comment (pengwa) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs:48]: /// A pointer to a underlying native instance of OrtTrainingSession

--> 

 /// A pointer to an underlying native instance of OrtTrainingSession

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs [hunk 8]
```diff
@@ -308,7 +401,14 @@ public float GetLearningRate()
         }
 
         /// <summary>
-        /// Runs a LR scheduler step. There must be a valid LR scheduler registered for the training session.
+        /// Update the learning rate based on the registered learing rate scheduler.
```
Comment (pengwa) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs:404]:
```suggestion
        /// Update the learning rate based on the registered learning rate scheduler.
```

File: docs/python/on_device_training/overview.rst [hunk 9]
```diff
@@ -1,7 +1,7 @@
 Overview
 =========
 
-Learning on the edge refers to the process of training a model on an edge device, such as mobile phones, embedded devices, gaming consoles, web browsers, etc. This is in contrast to training a model on a server or a cloud. Training on the edge is useful when the data is sensitive and cannot be shared with a server or a cloud. It is also useful for the task of personalization where the model needs to be trained on the user's device.
+`On-Device Training` refers to the process of training a model on an edge device, such as mobile phones, embedded devices, gaming consoles, web browsers, etc. This is in contrast to training a model on a server or a cloud. Training on the edge is useful when the data is sensitive and cannot be shared with a server or a cloud. It is also useful for the task of personalization where the model needs to be trained on the user's device.
 
 `onnxruntime-training` offers an easy way to efficiently train and infer a wide range of ONNX models on edge devices. The training process is divided into two phases:
```
Comment (pengwa) [inline docs/python/on_device_training/overview.rst:6]: Once training on the edge device is complete, an inference ready onnx model can be generated on the edge device itself. This model can then be used for inferencing with onnxruntime.

--> 

Once training on the edge device is complete, an inference-ready onnx model can be generated on the edge device itself. This model can then be used for inferencing with ONNX Runtime.

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 10]
```diff
@@ -9,7 +9,7 @@
  *
  * ONNX Runtime's C, C++ APIs offer an easy to use interface to onboard and execute onnx models.
```
Comment (pengwa) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:10]: * ONNX Runtime is a high-performance inference and training graph execution engine for deeplearning models.
--> 

 * ONNX Runtime is a high-performance inference and training graph execution engine for deep learning models.

**Review by natke — APPROVED** [review]
:shipit:  🚀

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs [hunk 11]
```diff
@@ -73,40 +88,57 @@ public static CheckpointState LoadCheckpoint(string checkpointPath)
         }
 
         /// <summary>
-        /// Saves the checkpoint
-        /// <param name="checkpointPath"> absolute path to the checkpoint file.</param>
-        /// <param name="includeOptimizerState"> absolute path to the checkpoint file.</param>
+        /// Save the given state to a checkpoint directory on disk.
+        ///
+        /// This function serializes the provided checkpoint state to a directory on disk.
+        /// This checkpoint can later be loaded by invoking CheckpointState.LoadCheckpoint to resume
```
Comment (askhade) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs:94]: begin/resume training?

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs [hunk 12]
```diff
@@ -62,57 +71,86 @@ public class TrainingSession : IDisposable
...
+        /// to the device specified by the user through the session options (if necessary).
         /// </summary>
-        /// <param name="state">Model checkpoint loaded into <see cref="CheckpointState"/>.</param>
-        /// <param name="trainModelPath">Specify path to training model graph.</param>
-        /// <param name="evalModelPath">Specify path to eval model graph.</param>
-        /// <param name="optimizerModelPath">Specify path to optimizer model graph.</param>
+        /// <param name="state">Training states that the training session uses as a starting point for training.</param>
+        /// <param name="trainModelPath">Model to be used to perform training.</param>
+        /// <param name="evalModelPath">Model to be used to perform evaluation.</param>
+        /// <param name="optimizerModelPath">Model to be used to perform gradient descent.</param>
```
Comment (askhade) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs:84]: nit: to perform weight update?

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs [hunk 13]
```diff
@@ -62,57 +71,86 @@ public class TrainingSession : IDisposable
...
+        /// This constructor instantiates the training session based on the env and session options provided that can
+        /// begin or resume training from a given checkpoint state for the given onnx models.
+        /// The checkpoint state represents the parameters of the training session which will be moved
+        /// to the device specified by the user through the session options (if necessary).
         /// </summary>
-        /// <param name="state">Model checkpoint loaded into <see cref="CheckpointState"/>.</param>
-        /// <param name="trainModelPath">Specify path to training model graph.</param>
+        /// <param name="state">Training states that the training session uses as a starting point for training.</param>
+        /// <param name="trainModelPath">Model to be used to perform training.</param>
         public TrainingSession(CheckpointState state, string trainModelPath)
```
Comment (askhade) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs:116]: what can we use this API for? Dont we always need optimizer model?

Comment (baijumeswani) [reply csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs:116]: I agree with you. I am not sure why we have this. I can probably remove this function altogether.

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs [hunk 14]
```diff
@@ -55,9 +64,15 @@ private void AddPropertyImpl<T>(string propertyName, PropertyType propertyType,
         public override bool IsInvalid { get { return handle == IntPtr.Zero; } }
 
         /// <summary>
-        /// Loads Checkpoint state from path
+        /// Load a checkpoint state from a directory on disk into checkpoint_state.
+        ///
+        /// This function will parse a checkpoint directory, pull relevant files and load the training
+        /// state into the checkpoint_state. This checkpoint state can then be used to create the
+        /// training session by instantiating the TrainingSession. By doing so, the training
+        /// session will resume training from the given checkpoint state.
```
Comment (askhade) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs:72]: same as above begin/resume

**Review by askhade — APPROVED** [review]
LGTM

**Review by jchen351 — APPROVED** [review]

**Review by pengwa — APPROVED** [review]

**Pull request merged**